File: lec0_intro1/ReturningData.c
#include <stdio.h>
#include <stdlib.h>

// Function that creates an array and returns it
int* createArray(int size) {
    int *arr = malloc(size * sizeof(int));
    if (arr == NULL) {
        return NULL;
    }
    
    for (int i = 0; i < size; i++) {
        arr[i] = i * 2;
    }
    
    return arr;  // Caller MUST remember to free this!
}

// Common mistake: returning stack-allocated memory
int* dangerousFunction() {
    int arr[5] = {1, 2, 3, 4, 5};
    return arr;  // BUG! This memory will be invalid after function returns!
}

int main() {
    int *numbers = createArray(5);
    
    if (numbers == NULL) {
        printf("Allocation failed!\n");
        return 1;
    }
    
    printf("Array: ");
    for (int i = 0; i < 5; i++) {
        printf("%d ", numbers[i]);
    }
    printf("\n");
    
    // Caller's responsibility to free!
    free(numbers);
    
    // Using the dangerous function would cause undefined behavior
    // int *bad = dangerousFunction();  // Don't do this!
    
    return 0;
}

/*
RETURNING DATA IN C:
1. Must return heap-allocated memory (malloc) if returning arrays.
2. Stack memory (local arrays) becomes invalid after function returns.
3. Caller must know to free the returned memory - easy to forget!
4. No automatic tracking of who "owns" the memory.
5. Documentation must specify who is responsible for freeing.
*/

File: lec0_intro1/ReturningData.java
package lec1_intro1;

public class ReturningData {
    // Function that creates and returns an array
    public static int[] createArray(int size) {
        int[] arr = new int[size];
        
        for (int i = 0; i < size; i++) {
            arr[i] = i * 2;
        }
        
        return arr;  // Safe! Java manages the memory.
    }
    
    // This is also safe in Java
    public static int[] safeFunction() {
        int[] arr = {1, 2, 3, 4, 5};
        return arr;  // Perfectly fine! Java keeps this memory alive.
    }
    
    public static void main(String[] args) {
        int[] numbers = createArray(5);
        
        System.out.print("Array: ");
        for (int i = 0; i < 5; i++) {
            System.out.print(numbers[i] + " ");
        }
        System.out.println();
        
        // No need to worry about freeing memory!
        // Garbage collector will clean up when no longer needed.
        
        int[] safe = safeFunction();
        System.out.print("Safe array: ");
        for (int i = 0; i < 5; i++) {
            System.out.print(safe[i] + " ");
        }
        System.out.println();
    }
}

/*
RETURNING DATA IN JAVA:
1. Can return arrays without worrying about stack vs heap.
2. Java automatically manages memory lifetime.
3. No concept of "dangling pointers" or "use after free".
4. Garbage collector tracks what's still in use.
5. Much simpler mental model - focus on logic, not memory!

KEY TAKEAWAY:
In C, you constantly think: "Where is this memory? Who frees it?"
In Java, you just use the data - Java handles the rest!
*/
